core/utils_test.py

```diff
@@ -794,3 +795,144 @@ def test_url_open(self) -> None:
+            'December 12 06:42:12',
+            utils.get_human_readable_time_string(944980932342.38)
+        )
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            AssertionError, 'Time cannot be negative'):
+            utils.get_human_readable_time_string(-1.42)
+
+    def test_generate_new_session_id(self) -> None:
+        test_string = utils.generate_new_session_id()
+        self.assertEqual(24, len(test_string))
+        self.assertIsInstance(test_string, str)
+        test_string_utf8 = test_string.encode('utf-8')
+        test_string_utf16 = test_string.encode('utf-16')
+        self.assertEqual(
+            len(test_string), len(test_string_utf8))
+        self.assertNotEqual(
+            len(test_string), len(test_string_utf16))
+
+    def test_require_valid_name_with_incorrect_input(self) -> None:
+        """ADD TEST ESCAPE SEQUENCE CHARACTERS CHECK."""
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            utils.ValidationError,
+            'The length of the exploration title should be between 1 and 50 ' 'characters; received '):   # pylint: disable=line-too-long
+            utils.require_valid_name('', 'the exploration title')
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            utils.ValidationError,
+            'Names should not start or end with whitespace.'):
+            utils.require_valid_name(' 123\n', 'the exploration title')
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            utils.ValidationError,
+            'Adjacent whitespace in the exploration title should be collapsed.'):   # pylint: disable=line-too-long
+            utils.require_valid_name('1  23', 'the exploration title')
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            utils.ValidationError,
+            'Invalid character : in the exploration title: 1\n:23'):
+            utils.require_valid_name('1\n:23', 'the exploration title')
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            utils.ValidationError,
+            r'Invalid character \\n in the exploration title: 1\\n23'):
+            utils.require_valid_name('1\\n23', 'the exploration title')
+
+    def test_get_hex_color_for_category(self) -> None:
+        self.assertEqual(
+            utils.get_hex_color_for_category('Law'), '#538270')
+        self.assertEqual(
+            utils.get_hex_color_for_category('Quantum Physics'), '#a33f40')
+
+    def test_unescape_encoded_uri_component(self) -> None:
+        self.assertEqual(
+            utils.unescape_encoded_uri_component('/El%20Ni%C3%B1o/'),
+            '/El Niño/')
+
+    def test_compress_and_decompress_zlib(self) -> None:
+        string_instance = b'a' * 26
+        string_compressed = utils.compress_to_zlib(string_instance)
+        self.assertNotEqual(
+            sys.getsizeof(string_compressed),
+            sys.getsizeof(string_instance))
```

Comment:
I think this is equivalent to comparing the lengths of the strings because Python's going to use the same number of bytes (1) per character (plus some extra space for metadata):

```python
>>> strings = ['a' * i for i in range(10)]
>>> strings
['', 'a', 'aa', 'aaa', 'aaaa', 'aaaaa', 'aaaaaa', 'aaaaaaa', 'aaaaaaaa', 'aaaaaaaaa']
>>> sizes = [sys.getsizeof(s) for s in strings]
>>> sizes
[49, 50, 51, 52, 53, 54, 55, 56, 57, 58]
```

Reply:
It isn't the same as comparing the lengths of the string because how zlib works is that it takes repeating sequences and compresses them to take less space that is why I have taken the case of b'a'*26. If, on the contrary, we used some random unique sequence 'asoierofinea' zlib won't reduce the space taken by the string.
So we are comparing the compressed space with the original space taken by the string.

Reply:
A couple points:

1. The variables you're working with here are of type `bytes`, not `str`, so naming them "string" is confusing. That's why I used strings in my example code above instead of bytes.
2. You're making a distinction between the length of a sequence of bytes and the space consumed by that sequence, but those two are directly related. A sequence of n bytes is always going to take up n bytes of space (unless there's some fancy compression going on under the hood, which I don't think is the case in Python).

I think if you changed this to assert that the compressed string takes up less space (and not just an unequal amount of space) to the uncompressed string, the test would fail. That's because compression doesn't always succeed in compressing the input data. Most compression algorithms need to add a little bit of metadata to describe how they compressed the data, and on short inputs, this metadata can take up more space than the compression saves. In this case, I think your input string is too short to be effectively compressed. Here's a longer example:

```python
>>> original = b'a' * 100
>>> len(original)
100
>>> sys.getsizeof(original)
133
>>> compressed = zlib.compress(original)
>>> len(compressed)
12
>>> sys.getsizeof(compressed)
45
```

For more details, see the [DEFLATE algorithm](https://en.wikipedia.org/wiki/Deflate), which is the underlying compression algorithm used by zlib.

Reply:
1. Changed the name from string to byte
2. I understand that the compression algo will not always succeed. However, I had taken that into account and then used b'a'*26. I have changed the code to assertLess and it works for this scenario. Should I include a case when the compression algo will not succeed?

Reply:
Nope, I don't think you need to test the case where compression fails